### PR TITLE
MC PX4: add flight behavior config page

### PR DESCRIPTION
**Pull request by bkueng**
From https://github.com/PX4/PX4-Autopilot/pull/16744

- A different logo would be nice
- Tab is only shown when the new parameter exists

image

### CODE REVIEW

File: src/AutoPilotPlugins/PX4/PX4AutoPilotPlugin.cc [hunk 1]
```diff
@@ -94,6 +95,12 @@ const QVariantList& PX4AutoPilotPlugin::vehicleComponents(void)
                 _tuningComponent->setupTriggerSignals();
                 _components.append(QVariant::fromValue((VehicleComponent*)_tuningComponent));
 
+                if(_vehicle->parameterManager()->parameterExists(_vehicle->id(), "SYS_VEHICLE_RESP")) {
+                    _flightBehavior = new PX4FlightBehavior(_vehicle, this);
```
Comment (patrickelectric) [inline src/AutoPilotPlugins/PX4/PX4AutoPilotPlugin.cc:99]: You are creating a dangling pointer, be sure to delete it or use a smart pointer (such as QSharedPointer), the first option is the desired way to accomplish it IMO.

File: src/AutoPilotPlugins/PX4/PX4AutoPilotPlugin.cc [hunk 2]
```diff
@@ -94,6 +95,12 @@ const QVariantList& PX4AutoPilotPlugin::vehicleComponents(void)
...
+                    _flightBehavior->setupTriggerSignals();
+                    _components.append(QVariant::fromValue((VehicleComponent*)_flightBehavior));
```
Comment (patrickelectric) [inline src/AutoPilotPlugins/PX4/PX4AutoPilotPlugin.cc:101]: avoid C cast style, use static_cast, dynamic_cast or reinterpret_cast

File: src/AutoPilotPlugins/PX4/PX4FlightBehavior.cc [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+#include "PX4AutoPilotPlugin.h"
+#include "AirframeComponent.h"
+
+PX4FlightBehavior::PX4FlightBehavior(Vehicle* vehicle, AutoPilotPlugin* autopilot, QObject* parent)
+    : VehicleComponent(vehicle, autopilot, parent)
+    , _name(tr("Flight Behavior"))
+{
+}
+
+QString PX4FlightBehavior::name(void) const
```
Comment (patrickelectric) [inline src/AutoPilotPlugins/PX4/PX4FlightBehavior.cc:21]: Just a note, (void) is unnecessary in C++ code.

File: src/AutoPilotPlugins/PX4/PX4FlightBehavior.cc [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            qmlFile = "qrc:/qml/PX4FlightBehaviorCopter.qml";
+            break;
+        case MAV_TYPE_VTOL_DUOROTOR:
+        case MAV_TYPE_VTOL_QUADROTOR:
+        case MAV_TYPE_VTOL_TILTROTOR:
+        case MAV_TYPE_VTOL_RESERVED2:
+        case MAV_TYPE_VTOL_RESERVED3:
+        case MAV_TYPE_VTOL_RESERVED4:
+        case MAV_TYPE_VTOL_RESERVED5:
+            qmlFile = "";
```
Comment (patrickelectric) [inline src/AutoPilotPlugins/PX4/PX4FlightBehavior.cc:74]: This attribution is unnecessary.

File: src/AutoPilotPlugins/PX4/PX4FlightBehaviorCopter.qml [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+                            description:    qsTr("A higher value makes the vehicle react faster. Be aware that this affects braking as well, and a combination of slow responsiveness with high maximum velocity will lead to long braking distances.")
+                            param:          "SYS_VEHICLE_RESP"
+                            min:            0.01
+                            max:            1
+                            step:           0.01
+                        }
+                    }
+                }
+                QGCLabel {
+                    visible:            _sys_vehicle_resp ? (_sys_vehicle_resp.value > 0.8) : false
```
Comment (patrickelectric) [inline src/AutoPilotPlugins/PX4/PX4FlightBehaviorCopter.qml:78]: `_sys_vehicle_resp && (_sys_vehicle_resp.value > 0.8)` ?

File: src/AutoPilotPlugins/PX4/PX4FlightBehaviorCopter.qml [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+                QGCCheckBox {
+                    id:                 xyVelCheckbox
+                    text:               qsTr("Enable horizontal velocity slider (if enabled, individual velocity limit parameters are automatically set)")
+                    checked:            _mpc_xy_vel_all ? (_mpc_xy_vel_all.value >= 0) : false
+                    onClicked: {
+                        if (checked) {
+                            _mpc_xy_vel_all.value = Math.abs(_mpc_xy_vel_all.value)
+                        } else {
+                            _mpc_xy_vel_all.value = -Math.abs(_mpc_xy_vel_all.value)
+                        }
```
Comment (patrickelectric) [inline src/AutoPilotPlugins/PX4/PX4FlightBehaviorCopter.qml:96]: maybe:
`_mpc_xy_vel_all.value = Math.abs(_mpc_xy_vel_all.value) * (checked ? 1.0 : -1.0)`
or
```
const mult = checked ? 1.0 : -1.0;
_mpc_xy_vel_all.value = mult*Math.abs(_mpc_xy_vel_all.value)
```

**Review by patrickelectric — CHANGES_REQUESTED** [review]

File: src/AutoPilotPlugins/PX4/PX4AutoPilotPlugin.cc (hunk 1 above)
Comment (bkueng) [reply src/AutoPilotPlugins/PX4/PX4AutoPilotPlugin.cc]: Yes, memory leak, which affects all other allocations as well. Fixed.

File: src/AutoPilotPlugins/PX4/PX4AutoPilotPlugin.cc (hunk 2 above)
Comment (bkueng) [reply src/AutoPilotPlugins/PX4/PX4AutoPilotPlugin.cc]: I was following the file's convention, so only partially related to the PR. I updated all.

File: src/AutoPilotPlugins/PX4/PX4FlightBehavior.cc (hunk 3 above)
Comment (bkueng) [reply src/AutoPilotPlugins/PX4/PX4FlightBehavior.cc]: copy-paste, and I agree

File: src/AutoPilotPlugins/PX4/PX4FlightBehavior.cc (hunk 4 above)
Comment (bkueng) [reply src/AutoPilotPlugins/PX4/PX4FlightBehavior.cc:74]: I was leaving this in for future

File: src/AutoPilotPlugins/PX4/PX4FlightBehaviorCopter.qml (hunk 6 above)
Comment (bkueng) [reply src/AutoPilotPlugins/PX4/PX4FlightBehaviorCopter.qml:96]: I was considering that but didn't really see a benefit.

**Review by LorenzMeier — APPROVED** [review]

**Comment by LorenzMeier** [comment]
Thanks!

**Pull request merged**